src/ray/object_manager/plasma/client.cc

```diff
@@ -581,16 +586,33 @@ Status PlasmaClient::Impl::Release(const ObjectID &object_id) {
   if (!store_conn_) {
     return Status::OK();
   }
-  auto object_entry = objects_in_use_.find(object_id);
+  const auto object_entry = objects_in_use_.find(object_id);
   RAY_CHECK(object_entry != objects_in_use_.end());
 
   object_entry->second->count -= 1;
   RAY_CHECK(object_entry->second->count >= 0);
   // Check if the client is no longer using this object.
   if (object_entry->second->count == 0) {
+    // object_entry is invalidated in MarkObjectUnused, need to read the fd beforehand.
+    MEMFD_TYPE fd = object_entry->second->object.store_fd;
     // Tell the store that the client no longer needs the object.
     RAY_RETURN_NOT_OK(MarkObjectUnused(object_id));
     RAY_RETURN_NOT_OK(SendReleaseRequest(store_conn_, object_id));
+    std::vector<uint8_t> buffer;
+    RAY_RETURN_NOT_OK(
+        PlasmaReceive(store_conn_, MessageType::PlasmaReleaseReply, &buffer));
+    ObjectID released_object_id;
+    bool should_unmap;
+    RAY_RETURN_NOT_OK(ReadReleaseReply(
+        buffer.data(), buffer.size(), &released_object_id, &should_unmap));
+    if (should_unmap) {
```

Comment:
Can you add a comment this is always false if the mmap is not fallback allocated? (it'd be nice to have assertion here, but I assume there's no way to check if mmap is fallback allocated vs not in client and add RAY_CHECK right?)

Reply:
yeah there's no way. In fact this is the main reason why we need to transfer this bool in the first place (because client can't tell)